src/coreclr/jit/codegenxarch.cpp

```diff
@@ -7336,7 +7336,20 @@ void CodeGen::genIntToFloatCast(GenTree* treeNode)
     // Also we don't expect to see uint32 -> float/double and uint64 -> float conversions
     // here since they should have been lowered appropriately.
     noway_assert(srcType != TYP_UINT);
-    noway_assert((srcType != TYP_ULONG) || (dstType != TYP_FLOAT));
+    assert((srcType != TYP_ULONG) || (dstType != TYP_FLOAT) ||
+           compiler->compIsaSupportedDebugOnly(InstructionSet_AVX512F));
+
+    if (compiler->compOpportunisticallyDependsOn(InstructionSet_AVX512F))
+    {
+        if (srcType == TYP_ULONG && (dstType == TYP_DOUBLE || dstType == TYP_FLOAT))
+        {
+            genConsumeOperands(treeNode->AsOp());
+            instruction ins = ins_FloatConv(dstType, srcType, emitTypeSize(srcType));
+            GetEmitter()->emitInsBinary(ins, emitTypeSize(srcType), treeNode, op1);
+            genProduceReg(treeNode);
+            return;
+        }
+    }
```

Comment:
It's better to use `varTypeIsFloating(dstType)`. While we're touching this, it gets a lot simply to have it be something like this instead:
```suggestion

    if ((srcType == TYP_ULONG) && varTypeIsFloating(dstType))
    {
        assert(compiler->compIsaSupportedDebugOnly(InstructionSet_AVX512F));
            
        genConsumeOperands(treeNode->AsOp());
        instruction ins = ins_FloatConv(dstType, srcType, emitTypeSize(srcType));
        GetEmitter()->emitInsBinary(ins, emitTypeSize(srcType), treeNode, op1);
        genProduceReg(treeNode);

        return;
    }
```

Reply:
yeah this seems better. Will make this changes

Reply:
Seems like we cannot make this change sicne ulong to double already use to reach here in non avx512 cases. Hence we need the if check and cannot put an assert.